lamagraph-compiler/src/LamaML/Syntax.hs

```diff
@@ -0,0 +1,290 @@
+
+/constr-name/ ::= /capitalized-ident/
+
+/typeconstr-name/ ::= /lowercase-ident/
+
+/field-name/ ::= /lowercase-ident/
+@
+
+/Revise or \/ || \/ lor/
+
+/Do we need records?/
+-}
+
+{- $types
+@
+/typexpr/ ::= ' /ident/
+          | _
+          | __(__ /typexpr/ __)__
+          | /typexpr/ -> /typexpr/
+          | /typexpr/ { * /typexpr/}+
+          | /typeconstr-name/
+          | /typexpr/ /typeconstr-name/
+          | __(__ /typexpr/ {, /typexpr/} __)__ /typeconstr-name/
+@
+-}
+
+{- $constants
+@
+/constant/ ::= /integer-literal/
+           | /char-literal/
+           | /string-literal/
+           | /constr-name/
+           | false
+           | true
+           | __()__
+           | begin end
+           | __[]__
+@
+-}
+
+{- $patterns
+@
+/pattern/ ::= /value-name/
+          | _
+          | /constant/
+          | __(__ /pattern/ __)__
+          | __(__ /pattern/ : /typexpr/ __)__
+          | /pattern/ __|__ /pattern/
+          | /constr-name/ /pattern/
+          | /pattern/ { , /pattern/ }
+          | __{__ /field-name/ [: /typexpr/] [= /pattern/] { ; /field-name/ [: /typexpr/] [= /pattern/] } [; _] [;] __}__
+          | __[__ /pattern/ { ; /pattern/ } [;] __]__
+          | /pattern/ :: /pattern/
+@
+-}
+
+{- $expressions
+
+@
+/expr/ ::= /value-name/
+       | /constant/
+       | __(__ /expr/ __)__
+       | begin /expr/ end
+       | /expr/ {, /expr/ }
+       | /constr-name/ /expr/
+       | /expr/ :: /expr/
+       | __[__ /expr/ { ; /expr/ } [;] ]
+       | __{__ /field-name/ [: /typexpr/] [= /expr/] { ; /field-name/ [: /typexpr/] [= /expr/] } [;] __}__
+       | __{__ /expr/ with /field-name/ [: /typexpr/] [= /expr/] { ; /field-name/ [: /typexpr/] [= /expr/] } [;] __}__
+       | /expr/ { /argument/ }+
+       | /prefix-symbol/ /expr/
+       | - /expr/
+       | /expr/ /infix-op/ /expr/
+       | if /expr/ then /expr/ else /expr/
+       | match /expr/ with /pattern-matching/
+       | function /pattern-matching/
+       | fun { /parameter/ }+ [ : /typexpr/ ] -> /expr/
+       | let [rec] /let-binding/ { and /let-binding/ } in /expr/
+
+/argument/ ::= /expr/
+
+/pattern-matching/ ::= [ __|__ ] /pattern/ [ when /expr/ ] -> /expr/ { __|__ /pattern/ [ when /expr/ ] -> /expr/ }
+
+/let-binding/ ::= /pattern/ = /expr/
+              | /value-name/ { /parameter/ } [ : /typexpr/ ] = /expr/
+
+/parameter/ ::= /pattern/
+@
+-}
+
+{- $typedefs
+
+@
+/type-definition/ ::= type /typedef/ { and /typedef/ }
+
+/typedef/ ::= [ /type-params/ ] /typeconstr-name/ /type-information/
+
+/type-information/ ::= [ /type-equation/ ] [ /type-representation/ ]
+
+/type-equation/ ::= = /typexpr/
+
+/type-representation/ ::= = [ __|__ ] /constr-decl/ { __|__ /constr-decl/ }
+                      | = /record=decl/
+                      | = __|__
+
+/type-params/ ::= /type-param/
+              | __(__ /type-param/ {, /type-param/ __)__
+
+/type-param/ ::= ' /ident/
+
+/record-decl/ ::= __{__ /field-decl/ { ; /field-decl/ } [;] __}__
```

Comment:
Do we plan to support records?

Reply:
Let's reduce complexity and forget about them, at least for now :)